prow/gerrit/gerrit.go

```diff
@@ -0,0 +1,189 @@
+*/
+
+// Package gerrit implements a gerrit-fetcher using https://github.com/andygrunwald/go-gerrit
+package gerrit
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/andygrunwald/go-gerrit"
+	"github.com/sirupsen/logrus"
+
+	"k8s.io/test-infra/prow/config"
+	"k8s.io/test-infra/prow/kube"
+	"k8s.io/test-infra/prow/pjutil"
+)
+
+type kubeClient interface {
+	CreateProwJob(kube.ProwJob) (kube.ProwJob, error)
+}
+
+type gerritClient interface {
+	QueryChanges(opt *gerrit.QueryChangeOptions) (*[]gerrit.ChangeInfo, *gerrit.Response, error)
+}
+
+type configAgent interface {
+	Config() *config.Config
+}
+
+// Controller manages gerrit changes.
+type Controller struct {
+	log *logrus.Entry
+	ca  configAgent
+
+	// go-gerrit change endpoint client
+	gc       gerritClient
+	instance string
+	projects []string
+
+	kc kubeClient
+
+	lastUpdate time.Time
+}
+
+// NewController returns a new gerrit controller client
+func NewController(instance string, projects []string, kc *kube.Client, logger *logrus.Entry, ca *config.Agent) (*Controller, error) {
+	c, err := gerrit.NewClient(instance, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	if logger == nil {
+		logger = logrus.NewEntry(logrus.StandardLogger())
+	}
+
+	return &Controller{
+		instance:   instance,
+		projects:   projects,
+		kc:         kc,
+		log:        logger,
+		ca:         ca,
+		gc:         c.Changes,
+		lastUpdate: time.Now(),
+	}, nil
+}
+
+// sync looks for newly made gerrit changes
+// and creates prowjobs according to presubmit specs
+func (c *Controller) Sync(rateLimit int) error {
+	syncTime := time.Now()
+	changes, err := c.QueryChanges(rateLimit)
+	if err != nil {
+		return fmt.Errorf("failed query changes : %v", err)
+	}
+
+	for _, change := range changes {
+		if err := c.ProcessChange(change); err != nil {
+			c.log.WithError(err).Error("Failed process change %v", change.CurrentRevision)
+		}
+	}
+
+	c.lastUpdate = syncTime
+	c.log.Infof("Processed %d changes", len(changes))
+	return nil
+}
+
+// QueryChanges will query all gerrit changes since controller's last sync loop
+func (c *Controller) QueryChanges(rateLimit int) (map[string]gerrit.ChangeInfo, error) {
+	opt := &gerrit.QueryChangeOptions{}
+	opt.Query = []string{}
+	for _, proj := range c.projects {
+		opt.Query = append(opt.Query, "project:"+proj)
+	}
+	opt.AdditionalFields = []string{"CURRENT_REVISION"}
+
+	done := false
+
+	// store a map of changeID:change
+	pending := map[string]gerrit.ChangeInfo{}
+	start := 0
+
+	for !done {
+		opt.Limit = rateLimit
+		opt.Start = start
+
+		// The change output is sorted by the last update time, most recently updated to oldest updated.
+		// Gerrit API docs: https://gerrit-review.googlesource.com/Documentation/rest-api-changes.html#list-changes
+		changes, _, err := c.gc.QueryChanges(opt)
+		if err != nil {
+			// should not happen? Let next sync loop catch up
+			return pending, fmt.Errorf("failed to query gerrit changes: %v", err)
+		}
+
+		if len(*changes) == 0 {
+			break
+		}
+
```

Comment:
`start` is incremented once for each change (all paths in the loop except the return path increment `start`) so you can remove the `start++` statements and put a single `start += len(changes)` here.